exporters/pelorus/deserialization/__init__.py

```diff
@@ -0,0 +1,392 @@
+
+
+# region: generic types
+
+# generic types such as dict[str, int] have two parts:
+# the "origin" and the "args".
+# The origin in this case is `dict`, and the args are `(str, int)`.
+# This lets us inspect these types properly.
+
+# in the future, these should be made into `typing.TypeGuard`s.
+
+
+def _is_dict_type(type_: type) -> Optional[tuple[type, type]]:
+    """
+    If this type annotation is a dictionary-like,
+    extract its key and value types.
+    Otherwise return None.
+    """
+    origin, args = typing.get_origin(type_), typing.get_args(type_)
+    if origin is None:
+        return None
+
+    if not issubclass(origin, MutableMapping):
+        return None
+
+    return args[0], args[1]
+
+
+def _is_list_type(type_: type) -> Optional[type]:
+    """
+    If this type annotation is a list-like,
+    extract its value type.
+    Otherwise return None.
+    """
+    origin, args = typing.get_origin(type_), typing.get_args(type_)
+    if origin is None:
+        return None
+
+    if not issubclass(origin, MutableSequence):
+        return None
+
+    return args[0]
+
+
+def _is_optional(type_: type) -> Optional[type]:
+    """
+    If this type annotation is Optional[T], then returns the type T.
+    Otherwise, returns None.
+    """
+    # this is weird because Optional is just an alias for Union,
+    # where the other entry is NoneType.
+
+    # it's unclear if order is guaranteed so we have to check both.
+    if typing.get_origin(type_) is not typing.Union:
+        return None
+
+    args = typing.get_args(type_)
+
+    if len(args) != 2:
+        return None
+
+    t1, t2 = args
+
+    if issubclass(t1, type(None)):  # type: ignore
+        return t2
+    elif issubclass(t2, type(None)):  # type: ignore
+        return t1
+    else:
+        return None
+
+
+# endregion
+
+
+@attrs.define
+class _Deserializer:
+    """
+    A deserializer that will collect errors in parallel.
+    """
+
+    unstructured_data_path: list[str] = attrs.field(factory=list, init=False)
+    "Where we are within the unstructured data."
+    structured_field_name_path: list[str] = attrs.field(factory=list, init=False)
+    "Where we are within the structured data."
+
+    def deserialize(
+        self, src: Any, target_type: type[T], src_name: str = "", target_name: str = ""
+    ) -> T:
+        """
+        Deserialize the unstructured src data into the given target type.
+        src_name and target_name are used for helpful error messages,
+        and will default to the names of the argument types.
+        """
+        target_name = target_name if target_name else str(target_type)
+
+        src_name = src_name if src_name else type(src).__name__
+
+        self.structured_field_name_path.append(target_name)
+        self.unstructured_data_path.append(src_name)
+        try:
+            return self._deserialize(src, target_type)
+        finally:
+            self.unstructured_data_path.pop()
```

Comment:
.pop() gets exactly one element, could you explain how this works if `self.unstructured_data_path` contains more then one element? I don't see recursion to ensure all elements are removed from this stack.